Server-errors-and-solutions.md

```diff
@@ -0,0 +1,84 @@
+
+### Preventive action
+
+1. __Design stage__: During design doc reviews, one should ask questions about performance and scalability when dealing with large amounts of data – "how big do you expect X to get?", "are there any limits to the size?", "what happens when it reaches this limit and is it possible to handle it?".
+2. __Manual testing__: Having some functionality to generate dummy data for testing locally might help.
+
+## Corner cases
+
+User-facing issues where certain corner case scenarios are not considered (e.g. empty list or undefined object). Example: Deleting a hint that is active in the editor view caused console errors. Although these are not often serious, these issues can be tricky to debug due to lack of reproduction steps. The majority of server errors fall under this category. Other examples of errors include: [#7907](https://github.com/oppia/oppia/pull/7907), [#7342](https://github.com/oppia/oppia/pull/7342), [#6523](https://github.com/oppia/oppia/pull/6523) (didn't account for changes to interaction type), [#5457](https://github.com/oppia/oppia/pull/5457).
+
+### Preventive action
+
+1. __Static analysis / compilation stage__: Having strong typing in the frontend will help catch some of these issues.
+2. __Manual testing__: Developers generally test only the "happy" paths in their feature. We should encourage them to test edge cases. Adoption of template users in testing mentioned [here](https://docs.google.com/document/d/1ktMagEKvUA6A0UmM_2CoENZPelqmjLmRjV7tWeo6w54/edit#heading=h.y3evobat23c5) should help uncover / test edge case scenarios.
+
+## Supporting old browsers
+
+Certain JS functionality may not be supported in older browsers. Example: String.prototype.match was used in the codebase but it lacked support in IE browsers. Fixes for these issues often involve adding a polyfill or swapping the relevant JS method with a more compatible one. Other examples of errors include: [#6909](https://github.com/oppia/oppia/pull/6909), [#6174](https://github.com/oppia/oppia/pull/6174), [#7926](https://github.com/oppia/oppia/pull/7926).
+
+### Preventive action
+
+1. __Static analysis / compilation stage__: Fix a JS standard that we expect all clients to support and specify a "target version" in typescript so that the code will be compiled to that version / JS standard.
+
+## Logic errors
+Logical errors in the code. Example: The "this" scope was referring to the global scope when it was supposed to refer to a local one ([#7873](https://github.com/oppia/oppia/pull/7873)). These are quite rare. Other examples of errors include: [#7872](https://github.com/oppia/oppia/issues/7872)(incorrect scope of "this"), [#7927](https://github.com/oppia/oppia/pull/7927).
```

Comment:
```suggestion
Logical errors in the code. Example: The "this" scope was referring to the global scope when it was supposed to refer to a local one ([#7873](https://github.com/oppia/oppia/pull/7873)). These are quite rare. Other examples of errors include: [#7872](https://github.com/oppia/oppia/issues/7872) (incorrect scope of "this"), [#7927](https://github.com/oppia/oppia/pull/7927).
```

Please proofread the rendered version to catch typos like this

Reply:
but their shouldn't be a gap between issue no. and bracket right?  I mean in general we write it as for eg.
28(twenty eight). So should I change it?

Reply:
There should be a gap between the issue number and parenthesis, and your example should be written as "28 (twenty eight)" with a space.

The only time I can think of where there wouldn't be a space before a parenthesis would be if you had a hierarchy of sections in a document. For example, if a question on an exam had a question number 1 with a subpart a, you could refer to the subpart as 1(a).

Reply:
done